src/tdlib/chat_list.rs

```diff
@@ -110,105 +92,91 @@ impl ChatList {
         }));
     }
 
-    pub(crate) fn handle_update(&self, update: Update) {
-        use Update::*;
-
-        match update {
-            ChatAction(ref update_) => self.handle_chat_update(update_.chat_id, update),
-            ChatDraftMessage(ref update_) => self.handle_chat_update(update_.chat_id, update),
-            ChatIsBlocked(ref update_) => self.handle_chat_update(update_.chat_id, update),
-            ChatIsMarkedAsUnread(ref update_) => self.handle_chat_update(update_.chat_id, update),
-            ChatLastMessage(ref update_) => self.handle_chat_update(update_.chat_id, update),
-            ChatNotificationSettings(ref update_) => {
-                self.handle_chat_update(update_.chat_id, update);
+    pub(crate) fn update_chat_position(&self, chat: &Chat, position: &TdChatPosition) {
```

Comment:
Uh, this doesn't look fun. This method is hardly readable because there's so much deep branching going on. I think the problem of keeping the order isn't solvable in a satisfying optimal way with the available data structures. On the one hand, we want to have fast lookup and on the other hand, fast insertion. But I think we can't have both without complicating things.

As a suggestion, I've replaced the `IndexMap`  with an `BTreeMap`. This eliminates much of the if-else branching and makes the code shorter and IMHO much more readable and maintainable. The disadvantage of the `BTreeMap` is the chat lookup complexity that is O(n). But I think this will not be significant, as I don't assume chat lists will become that large that we would see an actual effect. The true advantage of the `BTeeMap` is moving elements (removing and inserting) in contrast to `IndexMap::move` and `IndexMap::shift_remove_full` which are O(n) according to the docs.

Remove ".zip" ending
[btreemap.patch.zip](https://github.com/melix99/telegrand/files/10678601/btreemap.patch.zip)

Reply:
Yeah, you're right, `BTreeMap` is much better to use in this case. Also, having fast lookup is not that much of a necessity anymore because most of the updates are now being managed by the `HashMap` in the `Session`.

Thanks a lot for the patch too! I'll add you as a co-author :)

Reply:
I think you could also insert the order into the map with the signs reversed. Then you could avoid the `rev()`.

Reply:
Right, I'll do that too.

Reply:
I also removed the `ChatListItem::order` property since it was unused.